### PR TITLE
[FLINK-25447][table-planner] Fix batch query cannot generate plan when a sorted view into multi sinks

### PR DESCRIPTION
**Pull request by swuferhong**
## What is the purpose of the change

This PR is aims to fix cannot generate plan error when insert a sorted view into multi sinks. The  error is caused by the sort information in traitSet do not pass to `IntermediateRelTable` when block optimize.
So I override `getCollationList` method in `IntermediateRelTable` to get new traitSet include sort information.

## Brief change log

- Override `getCollationList` method in `IntermediateRelTable` to get new traitSet include sort information.
- Adding UT test for this case.
- Adding ITCase for this case.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no docs.

### CODE REVIEW

**Comment by swuferhong** [comment]
@godfreyhe @lincoln-lil  Could you review this pr ? Thanks !

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SortTest.scala [hunk 1]
```diff
@@ -46,6 +46,134 @@ class SortTest extends TableTestBase {
...
+                     |""".stripMargin)
+    val stmtSet = env.createStatementSet()
+    stmtSet.addInsertSql("INSERT INTO sink1 select a, b, c from Source order by d")
+    stmtSet.addInsertSql("INSERT INTO sink2 select a, b, c, d from Source order by d")
+
+    util.verifyExecPlan(stmtSet)
+  }
+
+  @Test
+  def testSortedViewResultIntoMultiSinks(): Unit = {
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SortTest.scala:95]: I think the test is more related to the {@link SubPlanReuseTest}. Could you move the test?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SortTest.scala [hunk 2]
```diff
@@ -46,6 +46,134 @@ class SortTest extends TableTestBase {
...
+                     |   d string
+                     |) WITH (
+                     |   'connector' = 'filesystem',
+                     |   'format' = 'testcsv',
+                     |   'path' = '/tmp/test'
+                     |)
+                     |""".stripMargin)
+    val stmtSet = env.createStatementSet()
+    stmtSet.addInsertSql("INSERT INTO sink2 select a, b, d, e from SortedTable")
+    util.verifyExecPlan(stmtSet)
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SortTest.scala:174]: This test is not related to the bug. I think we can remove this.

**Review by fsk119 — COMMENTED** [review]
Thanks for your contribution. The PR looks good in general. But I have the question why it works after providing collationList. I think rules are enough to provide the conversition.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SortTest.scala (hunk 1 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SortTest.scala]: > I think the test is more related to the {@link SubPlanReuseTest}. Could you move the test?

Done!

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SortTest.scala (hunk 2 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SortTest.scala]: > This test is not related to the bug. I think we can remove this.

Done!

**Comment by fsk119** [comment]
@flinkbot run azure

**Review by fsk119 — COMMENTED** [review]
Thanks for your update. I left some comments.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala [hunk 3]
```diff
@@ -334,6 +334,44 @@ class SubplanReuseTest extends TableTestBase {
     util.verifyExecPlan(sqlQuery)
   }
 
+  @Test
+  def testSubplanReuseOnSortedView(): Unit = {
+    util.addTableSource[(Int, Long, String, String, String)]("Source", 'a, 'b, 'c, 'd, 'e)
+    val query = "SELECT * FROM Source order by d"
+    val table = util.tableEnv.sqlQuery(query)
+    // Define a sorted view.
+    util.tableEnv.registerTable("SortedTable", table)
```
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala:343]: > use `createTemporaryView`. It's better we don't use deprecated method in the future.

Done!

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/SubplanReuseTest.scala [hunk 4]
```diff
@@ -248,6 +248,44 @@ class SubplanReuseTest extends TableTestBase {
     util.verifyExecPlan(sqlQuery)
   }
 
+  @Test
+  def testSubplanReuseOnSortedView(): Unit = {
+    util.addTableSource[(Int, Long, String, String, String)]("Source", 'a, 'b, 'c, 'd, 'e)
+    val query = "SELECT * FROM Source order by d"
+    val table = util.tableEnv.sqlQuery(query)
+    // Define a sorted view.
+    util.tableEnv.registerTable("SortedTable", table)
```
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/SubplanReuseTest.scala:257]: > ditto

This problem will not happen in stream mode. I removed this test.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdDistribution.scala [hunk 5]
```diff
@@ -87,6 +87,10 @@ class FlinkRelMdDistribution private extends MetadataHandler[FlinkDistribution]
     }
   }
 
+  def flinkDistribution(exchange: Exchange, mq: RelMetadataQuery): FlinkRelDistribution = {
+    exchange.distribution.asInstanceOf[FlinkRelDistribution]
```
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdDistribution.scala:91]: > It's not safe to just cast to FlinkRelDistribution. I think we should
> 
> ```
> // if exchagne.distribution is instance of FlinkRelDistribution 
> // return exchange.distribution
> 
> // else convert to FlinkRelDistribution according to RelDistribution.Type 
> ```

The RelNode call this method `flinkDistribution()` must be an instance of `FlinkRelDistribution`

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/schema/IntermediateRelTable.scala [hunk 6]
```diff
@@ -54,4 +55,19 @@ class IntermediateRelTable(
...
     this(names, relNode, ModifyKindSet.INSERT_ONLY, false, new util.HashSet[ImmutableBitSet]())
   }
+
+  override def getCollationList: util.List[RelCollation] = {
+    val mq = relNode.getCluster.getMetadataQuery
+    mq.collations(relNode)
+  }
+
+  override def getDistribution: RelDistribution = {
+    val fmq = FlinkRelMetadataQuery.reuseOrCreate(relNode.getCluster.getMetadataQuery)
```
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/schema/IntermediateRelTable.scala:65]: > Why don't use
> 
> ```
>     val fmq = relNode.getCluster.getMetadataQuery
>     fmq.distribution(relNode)
> ```

For `RelDistribution`, we must convert `RelMetadataQuery` to `FlinkRelMetadataQuery`, and call `fmq.flinkDistribution(relNode)` instead of `fmq.distribution(relNode)` to get `RelDistribution` (Here is a bit confusing. I still need to have a good understand).

[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/schema/IntermediateRelTable.scala:65]: > Why don't use
> 
> ```
>     val fmq = relNode.getCluster.getMetadataQuery
>     fmq.distribution(relNode)
> ```

The implementation of FlinkDistribution and Calcite Distribution is different, and theoretically they need to be the same. I can create another PR to unify them.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdDistribution.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdDistribution.scala]: > It's not safe to just cast to FlinkRelDistribution. I think we should
> 
> ```
> // if exchagne.distribution is instance of FlinkRelDistribution 
> // return exchange.distribution
> 
> // else convert to FlinkRelDistribution according to RelDistribution.Type 
> ```

Done!

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnUniquenessTest.scala [hunk 7]
```diff
@@ -54,6 +60,45 @@ class FlinkRelMdColumnUniquenessTest extends FlinkRelMdHandlerTestBase {
...
+    relBuilder.clear()
+    val studentRel = relBuilder
+      .scan("student")
+      .build()
+    val flinkLogicalIntermediateTableScan: FlinkLogicalIntermediateTableScan =
+      createIntermediateScan(
+        studentRel,
+        flinkLogicalTraits,
+        Set(ImmutableBitSet.of(0)),
+        new FlinkStatistic(new TableStats(1000L), Set(Set("name").asJava).asJava))
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnUniquenessTest.scala:74]: I think it should be 
```

 createIntermediateScan(
        studentRel,
        flinkLogicalTraits,
        Set(ImmutableBitSet.of(1)),
        new FlinkStatistic(new TableStats(1000L), Set(Set("name").asJava).asJava))
);
```

Because the current student schema is 
```
    val schema = new TableSchema(
      Array("id", "name", "score", "age", "height", "sex", "class"),
      Array(
        BasicTypeInfo.LONG_TYPE_INFO,
        BasicTypeInfo.STRING_TYPE_INFO,
        BasicTypeInfo.DOUBLE_TYPE_INFO,
        BasicTypeInfo.INT_TYPE_INFO,
        BasicTypeInfo.DOUBLE_TYPE_INFO,
        BasicTypeInfo.STRING_TYPE_INFO,
        BasicTypeInfo.INT_TYPE_INFO
      ))
```

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnUniquenessTest.scala [hunk 8]
```diff
@@ -54,6 +60,45 @@ class FlinkRelMdColumnUniquenessTest extends FlinkRelMdHandlerTestBase {
...
+        streamPhysicalTraits,
+        Set(ImmutableBitSet.of(0)),
+        new FlinkStatistic(new TableStats(1000L), Set(Set("name").asJava).asJava))
+    Array(
+      flinkLogicalIntermediateTableScan,
+      batchPhysicalIntermediateTableScan,
+      streamPhysicalIntermediateTableScan).foreach {
+      scan =>
+        assertFalse(mq.areColumnsUnique(scan, ImmutableBitSet.of()))
+        assertTrue(mq.areColumnsUnique(scan, ImmutableBitSet.of(0)))
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnUniquenessTest.scala:93]: It's very strange. If the intermediate table has the upsert key, I think we should respect this. 

In this case, if you treat the column `name` as the upsert keys, I think `id` should not be unique. For example, the intermediate table represents the query

```
SELECT 1 AS id, name FROM STUDENTS GROUP BY name
```

In this case, I think we can not have the promise.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala [hunk 9]
```diff
@@ -334,6 +334,43 @@ class SubplanReuseTest extends TableTestBase {
...
+    val table = util.tableEnv.sqlQuery(query)
+    // Define a sorted view.
+    util.tableEnv.createTemporaryView("SortedView", table)
+    util.tableEnv.executeSql("""
+                               |CREATE TABLE Sink1 (
+                               |   a int,
+                               |   b bigint,
+                               |   c string
+                               |) WITH (
+                               |   'connector' = 'filesystem',
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala:350]: nit: it's better we use values source/sink in the test.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala [hunk 10]
```diff
@@ -334,6 +334,43 @@ class SubplanReuseTest extends TableTestBase {
...
+                               |   c string,
+                               |   d string
+                               |) WITH (
+                               |   'connector' = 'filesystem',
+                               |   'format' = 'testcsv',
+                               |   'path' = '/tmp/test'
+                               |)
+                               |""".stripMargin)
+    val stmtSet = util.tableEnv.createStatementSet()
+    stmtSet.addInsertSql("INSERT INTO Sink1 select a, b, e from SortedView")
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala:368]: I think it's better to use the SQL in the jira to verify we have already fix this problem.

**Review by fsk119 — COMMENTED** [review]
Thanks for your update.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala (hunk 10 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala:368]: > I think it's better to use the SQL in the jira to verify we have already fix this problem.

Done!

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala (hunk 9 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SubplanReuseTest.scala:350]: > nit: it's better we use values source/sink in the test.

Done!

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnUniquenessTest.scala (hunk 7 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnUniquenessTest.scala]: I will remove this code.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnUniquenessTest.scala (hunk 8 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnUniquenessTest.scala]: > It's very strange. If the intermediate table has the upsert key, I think we should respect this.
> 
> In this case, if you treat the column `name` as the upsert keys, I think `id` should not be unique. For example, the intermediate table represents the query
> 
> ```
> SELECT 1 AS id, name FROM STUDENTS GROUP BY name
> ```
> 
> In this case, I think we can not have the promise.

I agree with you. We need not to override `FlinkPreparingTableBase.getKeys()`.

**Review by fsk119 — APPROVED** [review]
LGTM
